Source/WebCore/xml/XMLHttpRequest.cpp

```diff
@@ -889,18 +889,28 @@ String XMLHttpRequest::statusText() const
     return m_response.httpStatusText();
 }
 
-void XMLHttpRequest::didFail(const ResourceError& error)
+void XMLHttpRequest::handleCancellation(const ResourceError&)
 {
-    Ref protectedThis { *this };
+    m_exceptionCode = AbortError;
+    queueTaskKeepingObjectAlive(*this, TaskSource::Networking, CancellableTask(m_abortErrorGroup, [this] {
+        abortError();
+    }));
+}
 
+void XMLHttpRequest::didFail(const ResourceError& error)
+{
     // If we are already in an error state, for instance we called abort(), bail out early.
     if (m_error)
         return;
 
+    if (auto* document = this->document()) {
+        if (auto* window = document->domWindow())
+            m_wasAbortedByClient |= window->isStopping();
+    }
+
     // The XHR specification says we should only fire an abort event if the cancelation was requested by the client.
     if (m_wasAbortedByClient && error.isCancellation()) {
```

Comment:
With this patch, m_wasAbortedByClient is always false, except in this method, and it is the only place we check it. Can we use a local variable?

Reply:
While it’s non-obvious, I studied the code and verified that @youennf is right about this. No other code is looking at `m_wasAbortedByClient`, and it will always be false entering this function, so we don’t need `m_wasAbortedByClient` any more.

Reply:
It is set to true in abort().

Reply:
But `abort` then unconditionally calls `immediateAbort`. That sets `m_error` unconditionally. So we will return early in that case and after the early return we are guaranteed that